### PR TITLE
Authentication of Oracle API endpoints

### PR DESCRIPTION
**Pull request by elijaholmos**
## summary
Implements API authentication for all Oracle endpoints. All non-GET methods require authentication. Furthermore, any attempts to modify existing data can only be performed by the owner of that data (the `oracleId` connected to the apikey must match the `oracleId` of the resource attempting to be modified)
A master API key is connected to `oracleId` 0 (a SYSTEM account) and has full access to all operations. It is the only allowed form of authorization for POSTing the `/oracle` endpoint.
API keys should be provided in the Authorization header of the HTTP request and follow the HTTP Bearer Authentication standard.

## todo
- Once the current implementation for Oracle endpoints is validated, I will apply a similar implementation for the other endpoints
- Expand the `AuthError` class and integrate it with the `ErrorResponse` model

### CODE REVIEW

**Comment by kharaone** [comment]
@elijaholmos I will need some help in order to reproduce your error and setup the whole api with auth

**Comment by kharaone** [comment]
@elijaholmos ok, it seems the error is tied to the connector : 
it's using `Object.prototype.hasOwnProperty.call(where, 'key')` but it's incorrectly setting the property name, instead of using the value contained in the variable, it's using the string name of the variable.
if I switch the `Object.prototype.hasOwnProperty.call(where, 'key')` to `Object.prototype.hasOwnProperty.call(where, key)` on line 460 of `src\connectors\BetterFirestoreConnector.ts` I'm getting through correctly.

feel free to review the connector code for these specific cases

**Review by kharaone — CHANGES_REQUESTED** [review]
see previous comment

**Comment by elijaholmos** [comment]
@kharaone I verified that your proposed solution fixed the bug - thank you for helping me solve it
PR description has been updated accordingly and is ready for the next stage of review

File: src/sequence.ts
Comment (kharaone) [inline src/sequence.ts]: I would have mentioned the fact that the middleware route seemed to be the correct way to go, but seeing you removed the ` extends MiddlewareSequence` makes me wonder why you chose to go back to the sequenceHandler ?

File: src/sequence.ts
Comment (kharaone) [inline src/sequence.ts]: This part is specific to /oracle endpoint and will need to be constrained accordingly. Moreover, it feels more like authorization than authentication, you will need to find a better place to define these rules

File: src/sequence.ts
Comment (kharaone) [inline src/sequence.ts]: needs to be awaited

File: src/sequence.ts
Comment (kharaone) [inline src/sequence.ts]: needs to be awaited

File: src/sequence.ts
Comment (kharaone) [inline src/sequence.ts]: if if's not a one liner, put the {} around the if scope

File: src/providers/verifyFn.provider.ts
Comment (kharaone) [inline src/providers/verifyFn.provider.ts]: Shouldn't we have a check on expired too ?

File: src/providers/verifyFn.provider.ts
Comment (kharaone) [inline src/providers/verifyFn.provider.ts]: to be removed

File: src/providers/verifyFn.provider.ts
Comment (kharaone) [inline src/providers/verifyFn.provider.ts]: to be removed

File: src/providers/verifyFn.provider.ts
Comment (kharaone) [inline src/providers/verifyFn.provider.ts]: to be removed

File: src/providers/verifyFn.provider.ts
Comment (kharaone) [inline src/providers/verifyFn.provider.ts:31]: if not a oneliner, please encapsulate with `{}`

File: src/oracleprofile.factory.ts
Comment (kharaone) [inline src/oracleprofile.factory.ts]:
```suggestion
  return {...oracle, [securityId]: oracle?.id?.toString() ?? ''} as GoodOracle;
```

File: src/oracleprofile.factory.ts
Comment (kharaone) [inline src/oracleprofile.factory.ts]: I'd move that file it in models

File: src/models/good-oracle.model.ts
Comment (kharaone) [inline src/models/good-oracle.model.ts]:
```suggestion
  [securityId]: string = this.id?.toString() ?? '';
```

File: src/providers/auth-error.ts
Comment (kharaone) [inline src/providers/auth-error.ts:1]: agreed, this should be translated into an errorresponse either through an interceptor or a middleware

File: src/providers/passport-bearer-auth.ts
Comment (kharaone) [inline src/providers/passport-bearer-auth.ts]: should probably contain `provider `in the filename to stay consistent in the naming convention

**Review by kharaone — CHANGES_REQUESTED** [review]
Here are a few suggestions on how to proceed further

**Review by kharaone — APPROVED** [review]
I'm still concerned of the part in sequence.ts which seemed specific to /oracle endpoint. You should at least guard the section with a conditional on the url 

if url === /oracle then apply rules.

**Comment by elijaholmos** [comment]
@kharaone sorry, that was only the first part 😅 I'll re-request review once I'm finished with all the suggested changes

File: src/sequence.ts
Comment (elijaholmos) [reply src/sequence.ts]: It was largely motivated by lb4's example found at https://github.com/loopbackio/loopback-next/blob/e2ff6b22367e919926d0f41f6d939d988c654c00/examples/todo-jwt/src/sequence.ts

File: src/sequence.ts
Comment (elijaholmos) [reply src/sequence.ts]: refactored in 8e71216344520cba30854a9c3ded48664ded42bd

**Comment by elijaholmos** [comment]
@kharaone two final questions:
1. for the sake of PR scope, could a separate story be created for the handling & displaying of API errors to the requester? including converting `AuthError`s into `ErrorResponse`s?
2. would you like me to implement authentication on the remaining endpoints before merging this? `/activity`, `/entry`, etc

**Comment by kharaone** [comment]
@elijaholmos 
1. sure, we can implement that in another story indeed
2. you should avoid /entry for now as @gitluky is getting that part overhauled. Feel free to either create separate stories for each endpoints or implement these in that PR

**Comment by elijaholmos** [comment]
@kharaone thank you, I will go ahead and move everything to separate stories so this can be merged once re-approved

**Review by kharaone — APPROVED** [review]
- I didn't update the tests with the authentication dimension (as there's apparently no way to insert an ID = 0 in the datasource, so we'll need to change the admin flag to be an actual flag)
- any person calling /auth with oracleId = X can "override" its apikey ... feels a bit problematic. We will probably need to do secret ids or even jwt token flow instead.

**Comment by elijaholmos** [comment]
I have a few ideas about the second point, for discussion at a future sync

**Pull request merged**
